### ISSUE TITLE
`CharacterBody3D` gets caught on ramp-to-flat boundary of `ConcavePolygonShape3D`

### ISSUE THREAD

**Issue by vfig** [godot-jolt/godot-jolt#524]
As [described on mastodon](https://mastodon.gamedev.place/@vfig/110810214294734819):

Often when moving from an upward sloping ramp part of a collider to the flat polygon at the top of the ramp, my character gets caught up. This issue wasn't seen before I moved over to godot-jolt.

A minimal repro project is attached. Additional notes with details on movement to repro the issue are in the readme in the project zip.

This is using Godot v4.1.stable.official [970459615] and godot-jolt v0.4.1-stable [061ac08] binary from github (excluded from repro zip for size).

[bug-ramp-caught-up.zip](https://github.com/godot-jolt/godot-jolt/files/12222632/bug-ramp-caught-up.zip)

**Comment by mihe** [comment]
Thank you for taking the time to create the MRP, and for writing up the notes in `readme.txt`. Much appreciated.

The repro for the stairs and planter were simple enough, thanks to your instructions, so I'll take a look at those after I've published 0.5.0-rc1 some time tomorrow.

I haven't been able to repro the crouching issue just yet, but considering your notes on the matter it might be worth mentioning that you can just disable the back-face collision for ray-casts as well, using the `hit_back_faces` parameter.

**Comment by vfig** [comment]
Ah, I had misread this bit from your readme, and took it to mean that rays would _necessarily_ hit the back faces, not just when the option was enabled. My bad.

> Ray-casts will hit the back-faces of all shape types, not just concave polygons and height maps

**Comment by mihe** [comment]
I'm not quite sure what to make of this one.

I tried replacing your `CharacterBody3D` script with the default template script in your project, and was still able to reproduce the issue at the top of the stairs, but when I then extracted the collision mesh for the stairs (i.e. the ramp) and put that into a new project, along with a `CharacterBody3D` with the exact same shape/properties, then I couldn't reproduce the issue anymore.

I can however reproduce the issue in this new project if I bump up the `safe_margin` property on the `CharacterBody3D` to something like 0.003-ish and above, as opposed to the default 0.001. What's interesting though is that Godot Physics behaves the exact same way with this value and also starts to get stuck at the top of the stairs.

Here's the new project I've been using: [Stairs.zip](https://github.com/godot-jolt/godot-jolt/files/12242009/Stairs.zip)

(You use the mouse scroll wheel to increase/decrease the safe margin.)

If you notice any obvious relevant discrepancy between that project and your own then I'd be curious to know what that is, because I can't seem to find it.

**Comment by vfig** [comment]
Huh, I didn't realise i had changed `floor_max_angle`. Still, the angle of the ramp is ~34°, so I wouldn't have expected changing floor_max_angle to have made a difference.

As it is happening with godot physics too (I had never encountered it there, but I guess that was just luck?) then yeah, seems like CharacterBody3D is implicated. I will take a look at your project too and see if I can see what's going on.

**Comment by mihe** [comment]
> Still, the angle of the ramp is ~34°, so I wouldn't have expected changing floor_max_angle to have made a difference.

Yeah, I didn't mean to imply that this breakage was somehow expected, but it does seem to be related none the less.

There seems to be some weird interplay between `safe_margin` and `floor_max_angle` where if you increase `safe_margin` then the floor angle (as seen by `move_and_slide`) is seemingly also increased, in the context of this particular boundary, leading to it thinking you're up against a wall. Similarly, if you decrease `safe_margin` then the inverse effect is observed.

I don't know if this is what's actually happening under the hood, but I'll see if I can step through the code to see what's going on. Maybe there's an obvious bug that can be resolved on the Godot side of things, since most of `move_and_slide` is implemented there.

**Comment by vfig** [comment]
Poking at your Stairs sample just now, though without being set up to step through godot's code, I am doing what I can with only post-mortem. Added the following around the `move_and_slide()` call to dump all the relevant details:
```
	var pre_position := position
	var pre_velocity := velocity
	move_and_slide()
	var post_position := position
	var post_velocity := velocity

	if not Input.is_action_pressed("char_forward"): return
	print("")
	print("Frame ",Engine.get_physics_frames())
	print("  pre-move position:",pre_position," velocity:",pre_velocity)
	print("  post-move position:",post_position," velocity:",post_velocity)
	print("  on floor:",is_on_floor()," only:",is_on_floor_only()," normal:",get_floor_normal()," angle:",rad_to_deg(get_floor_angle()))
	print("  on wall:",is_on_wall()," only:",is_on_wall_only()," normal:",get_wall_normal())
	print("  on ceiling:",is_on_ceiling()," only:",is_on_ceiling_only())
	print("Slide collisions:",get_slide_collision_count())
	for i in get_slide_collision_count():
		var sc := get_slide_collision(i)
		print("  ",i,". depth:",sc.get_depth()," travel:",sc.get_travel()," remainder:",sc.get_remainder()," count:",sc.get_collision_count())
		for j in sc.get_collision_count():
			print("  ",j,") position:",sc.get_position(j)," normal:",sc.get_normal(j)," angle:",rad_to_deg(sc.get_angle(j))," local shape:",sc.get_local_shape(j))
			print("      collider:",sc.get_collider(j)," shape:",sc.get_collider_shape(j)," velocity:",sc.get_collider_velocity(j))
	print("")
```

I enabled godot-jolt, and I set safe_margin to 0.003 because it makes the problem occur more often. Sure enough, got caught up on the simple ramp almost right away, exactly on the edge between the ramp and the flat:

![collision2pt](https://github.com/godot-jolt/godot-jolt/assets/81129/5f47b0f9-205c-4d26-8419-4f163e3859bd)

![collision2side](https://github.com/godot-jolt/godot-jolt/assets/81129/5b8497f0-1f8f-4945-b560-98b2c1685100)

The most obvious issue from the log is that it is reporting `is_on_wall()==true`. Slide collision 1 appears to be colliding with the _edge_ of the flat area (the reported normal does not make sense otherwise). And the normal it reports is at just steep enough an angle to be flagged as a wall.

```
Frame 502
  pre-move position:(-0.047784, 1.203981, -1.31119) velocity:(-0.020942, 0, -0.999781)
  post-move position:(-0.047784, 1.203981, -1.31119) velocity:(0, 0, 0)
  on floor:true only:false normal:(0, 0.832067, 0.554674) angle:33.6882927413492
  on wall:true only:false normal:(0, 0.687293, 0.726381)
  on ceiling:false only:false
Slide collisions:1
  0. depth:0.003513334319 travel:(-0.000105, 0, -0.00503) remainder:(-0.000244, 0, -0.011633) count:2
  0) position:(-0.047915, 0.984826, -1.383538) normal:(0, 0.832067, 0.554674) angle:33.6882927413492 local shape:Shape:<CollisionShape3D#28085060811>
      collider:Ramp:<StaticBody3D#27984397509> shape:Shape_Ramp:<CollisionShape3D#28001174726> velocity:(0, 0, 0)
  1) position:(-0.047915, 1, -1.4063) normal:(0, 0.687293, 0.726381) angle:46.5838333283058 local shape:Shape:<CollisionShape3D#28085060811>
      collider:Ramp:<StaticBody3D#27984397509> shape:Shape_Ramp:<CollisionShape3D#28001174726> velocity:(0, 0, 0)
```

I have looked around in the implementation of move_and_slide(), but there's a lot going on in there, and honestly I can't follow it well enough by just reading it. However just from all the above, I have two workarounds, either of which prevents the hitching:

- raise `floor_max_angle` a little, e.g. to 50 degrees
- or turn off `floor_block_on_wall`

(though I don't understand from the documentation what `floor_block_on_wall` is actually supposed to do ordinarly)

Anyway, why this 46 degree normal is being reported in the first place, I don't fully understand, probably because I don't understand how `margin` or `safe_margin` work in terms of the precision of the collision. However from just zooming in Real Close (tm) at the side view with the character at the position it would be in after the "travel" (if I am understanding correctly what that means), then I can see the edge is just _above_ the 45 degree point on the capsule bottom. I guess that explains the normal? I wonder if this only happens because `move_and_slide()` is setting `recovery_as_collision=true` when calling `move_and_collide()`?

![collision2superzoom](https://github.com/godot-jolt/godot-jolt/assets/81129/49de2bdc-6e7d-4375-b3b3-b78402527720)

**Comment by mihe** [comment]
It seems this issue was somewhat related to #508 all along, which I probably should have realized.

While this isn't 100% fixed, due to `move_and_slide` still reporting multiple slide collisions with varying normals (one of which is typically exaggarated), the effect seems to have been greatly lessened by #546 to the point where I'm comfortable closing this issue.

I'll try to have a 0.6.0-rc1 out some time tomorrow, so you can try the fix for yourself.

**Comment by mihe** [comment]
I've just published [0.6.0-rc1](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.6.0-rc1), which includes the above mentioned "fix". I'd appreciate it if you (@vfig) could give it a try and confirm that it does resolve the issue as far as your particular setup is concerned.

**Comment by vfig** [comment]
this does appear to have resolved my issue (0.6.0-stable)! no longer getting caught up on either of the problem cases.
